plutus-ir/src/Language/PlutusIR/Parser.hs

```diff
@@ -0,0 +1,344 @@
+    <|> TyBuiltin <$> reservedWord "integer" <*> return PLC.TyInteger
+    <|> TyBuiltin <$> reservedWord "bytestring" <*> return PLC.TyByteString
+    <|> TyInt <$> getPosition <*> Lex.decimal
+
+appType :: Parser (Type TyName SourcePos)
+appType = do
+    pos  <- getPosition
+    fn   <- typ
+    args <- some typ
+    pure $ foldl' (TyApp pos) fn args
+
+kind :: Parser (Kind SourcePos)
+kind = inParens (typeKind <|> sizeKind <|> funKind)
+    where
+        typeKind = Type <$> reservedWord "type"
+        sizeKind = Size <$> reservedWord "size"
+        funKind  = KindArrow <$> reservedWord "fun" <*> kind <*> kind
+
+typ :: Parser (Type TyName SourcePos)
+typ = (tyVar >>= (\n -> return $ TyVar (nameAttribute $ unTyName n) n))
+    <|> (inParens $ funType <|> allType <|> lamType <|> ifixType <|> conType)
+    <|> inBrackets appType
+
+varDecl :: Parser (VarDecl TyName Name SourcePos)
+varDecl = inParens $ VarDecl <$> reservedWord "vardecl" <*> var <*> typ
+
+tyVarDecl :: Parser (TyVarDecl TyName SourcePos)
+tyVarDecl = inParens $ TyVarDecl <$> reservedWord "tyvardecl" <*> tyVar <*> kind
+
+datatype :: Parser (Datatype TyName Name SourcePos)
+datatype = inParens $ Datatype <$> reservedWord "datatype"
+    <*> tyVarDecl
+    <*> some tyVarDecl
+    <*> var
+    <*> some varDecl
+
+binding :: Parser (Binding TyName Name SourcePos)
+binding =  inParens $
+    (try $ reservedWord "termbind" >> TermBind <$> getPosition <*> varDecl <*> term)
+    <|> (reservedWord "typebind" >> TypeBind <$> getPosition <*> tyVarDecl <*> typ)
+    <|> (reservedWord "datatypebind" >> DatatypeBind <$> getPosition <*> datatype)
+
+absTerm :: Parser (Term TyName Name SourcePos)
+absTerm = TyAbs <$> reservedWord "abs" <*> tyVar <*> kind <*> term
+
+lamTerm :: Parser (Term TyName Name SourcePos)
+lamTerm = LamAbs <$> reservedWord "lam" <*> name <*> typ <*> term
+
+conTerm :: Parser (Term TyName Name SourcePos)
+conTerm = Constant <$> reservedWord "con" <*> constant
+
+iwrapTerm :: Parser (Term TyName Name SourcePos)
+iwrapTerm = IWrap <$> reservedWord "iwrap" <*> typ <*> typ <*> term
+
+builtinTerm :: Parser (Term TyName Name SourcePos)
+builtinTerm = Builtin <$> reservedWord "builtin" <*> builtinVar
+
+unwrapTerm :: Parser (Term TyName Name SourcePos)
+unwrapTerm = Unwrap <$> reservedWord "unwrap" <*> term
+
+errorTerm :: Parser (Term TyName Name SourcePos)
+errorTerm = Error <$> reservedWord "error" <*> typ
+
+letTerm :: Parser (Term TyName Name SourcePos)
+letTerm = Let <$> reservedWord "let" <*> recursivity <*> some (try binding) <*> term
+
+appTerm :: Parser (Term TyName Name SourcePos)
+appTerm = do
+    pos  <- getPosition
+    fn   <- term
+    args <- some term
+    pure $ foldl' (Apply pos) fn args
+
+tyInstTerm :: Parser (Term TyName Name SourcePos)
+tyInstTerm = do
+    pos  <- getPosition
+    fn   <- term
+    args <- some typ
+    pure $ foldl' (TyInst pos) fn args
+
+term :: Parser (Term TyName Name SourcePos)
```

Comment:
I think we should take a slightly different approach, namely:
- Actually have a PLC parser in a different module.
- Use the `term` parser of the PLC parser as part of the PIR parser by applying `embedIntoIR` on it.

Reply:
This can be done, but it's a bit trickier since `term` is recursive so you can't really say "a PIR term is either a PLC term or a let" - I think it's only worth the extra complexity if we really plan to migrate the PLC parser to megaparsec at some point

Reply:
Ah that's true, we wouldn't get interleaving that way. Annoying.

Reply:
We can, and it's not super hard, just make the rule take an extra argument for "other stuff that you might want me to parse", the PLC parser just passes `void` and the PIR parser passes the rule for `let`, it's just a bit annoying - same goes for reusing the PLC generators. I think the investment might be worth it in the long run.

Reply:
Or, more elegantly, just write `term1` as one step of recursion, then `plcTerm = fix term1` and `pirTerm = fix ((letTerm <|>) . term1)` (I think)